File: apps/mobile/src/services/ml/features.ts
/**
 * Feature extraction for the on-device motion model.
 *
 * EXACT port of services/ml-training/features.py — the 24 features must be
 * produced in the same order with the same math, or a model trained in Python
 * will misbehave on device. Uses population std (ddof = 0) to match numpy.
 *
 * Input : window as number[][], shape (WINDOW_SIZE, 6) = [ax,ay,az,gx,gy,gz]
 * Output: Float array of length 24 (see FEATURE_NAMES order in features.py)
 */

function mean(a: number[]): number {
  let s = 0;
  for (const v of a) s += v;
  return a.length ? s / a.length : 0;
}

function std(a: number[], m?: number): number {
  if (a.length === 0) return 0;
  const mu = m ?? mean(a);
  let s = 0;
  for (const v of a) s += (v - mu) * (v - mu);
  return Math.sqrt(s / a.length); // population std (ddof=0), matches numpy
}

function min(a: number[]): number {
  let m = Infinity;
  for (const v of a) if (v < m) m = v;
  return m;
}

function max(a: number[]): number {
  let m = -Infinity;
  for (const v of a) if (v > m) m = v;
  return m;
}

function meanAbsDev(a: number[], m: number): number {
  let s = 0;
  for (const v of a) s += Math.abs(v - m);
  return a.length ? s / a.length : 0;
}

function meanSquares(a: number[]): number {
  let s = 0;
  for (const v of a) s += v * v;
  return a.length ? s / a.length : 0;
}

export function extractFeatures(window: number[][]): number[] {
  const n = window.length;
  const ax: number[] = new Array(n);
  const ay: number[] = new Array(n);
  const az: number[] = new Array(n);
  const gx: number[] = new Array(n);
  const gy: number[] = new Array(n);
  const gz: number[] = new Array(n);
  const accMag: number[] = new Array(n);
  const gyroMag: number[] = new Array(n);

  for (let i = 0; i < n; i++) {
    const r = window[i];
    ax[i] = r[0]; ay[i] = r[1]; az[i] = r[2];
    gx[i] = r[3]; gy[i] = r[4]; gz[i] = r[5];
    accMag[i] = Math.sqrt(r[0] * r[0] + r[1] * r[1] + r[2] * r[2]);
    gyroMag[i] = Math.sqrt(r[3] * r[3] + r[4] * r[4] + r[5] * r[5]);
  }

  // accel jerk = |diff(accMag)|
  const accJerk: number[] = [];
  for (let i = 1; i < n; i++) accJerk.push(Math.abs(accMag[i] - accMag[i - 1]));
  if (accJerk.length === 0) accJerk.push(0);

  const accMagMean = mean(accMag);

  // accel / gyro SMA = mean over samples of (|x|+|y|+|z|)
  let accSma = 0;
  let gyroSma = 0;
  for (let i = 0; i < n; i++) {
    accSma += Math.abs(ax[i]) + Math.abs(ay[i]) + Math.abs(az[i]);
    gyroSma += Math.abs(gx[i]) + Math.abs(gy[i]) + Math.abs(gz[i]);
  }
  accSma = n ? accSma / n : 0;
  gyroSma = n ? gyroSma / n : 0;

  return [
    accMagMean,
    std(accMag, accMagMean),
    min(accMag),
    max(accMag),
    max(accMag) - min(accMag),
    meanSquares(accMag),
    meanAbsDev(accMag, accMagMean),

    mean(accJerk),
    max(accJerk),

    std(ax), std(ay), std(az),
    mean(ax), mean(ay), mean(az),
    accSma,

    mean(gyroMag),
    std(gyroMag),
    max(gyroMag),
    meanSquares(gyroMag),

    std(gx), std(gy), std(gz),
    gyroSma,
  ];
}

File: services/ml-training/features.py
"""
Feature extraction for the motion model.

Input : a window of shape (WINDOW_SIZE, 6) — columns [ax, ay, az, gx, gy, gz].
Output: a fixed-length feature vector (see FEATURE_NAMES).

This is the SINGLE source of truth for the feature math. The TypeScript
port in apps/mobile/src/services/ml/features.ts must produce identical
values so that a model trained here behaves the same on device.
"""

import numpy as np

# Ordered feature names — the model input columns, in this exact order.
FEATURE_NAMES = [
    # accel magnitude stats
    "acc_mag_mean", "acc_mag_std", "acc_mag_min", "acc_mag_max",
    "acc_mag_range", "acc_mag_energy", "acc_mag_mad",
    # accel jerk (rate of change of magnitude)
    "acc_jerk_mean_abs", "acc_jerk_max_abs",
    # per-axis accel std (posture / orientation change)
    "ax_std", "ay_std", "az_std",
    # per-axis accel mean (gravity orientation)
    "ax_mean", "ay_mean", "az_mean",
    # signal magnitude area (accel)
    "acc_sma",
    # gyro magnitude stats
    "gyro_mag_mean", "gyro_mag_std", "gyro_mag_max", "gyro_mag_energy",
    # per-axis gyro std (rotational agitation)
    "gx_std", "gy_std", "gz_std",
    # gyro signal magnitude area
    "gyro_sma",
]
NUM_FEATURES = len(FEATURE_NAMES)


def extract_features(window: np.ndarray) -> np.ndarray:
    """window: (WINDOW_SIZE, 6) float array -> (NUM_FEATURES,) float array."""
    w = np.asarray(window, dtype=np.float64)
    ax, ay, az = w[:, 0], w[:, 1], w[:, 2]
    gx, gy, gz = w[:, 3], w[:, 4], w[:, 5]

    acc_mag = np.sqrt(ax * ax + ay * ay + az * az)
    gyro_mag = np.sqrt(gx * gx + gy * gy + gz * gz)

    acc_jerk = np.abs(np.diff(acc_mag)) if len(acc_mag) > 1 else np.array([0.0])

    feats = [
        acc_mag.mean(),
        acc_mag.std(),
        acc_mag.min(),
        acc_mag.max(),
        acc_mag.max() - acc_mag.min(),
        np.mean(acc_mag ** 2),
        np.mean(np.abs(acc_mag - acc_mag.mean())),  # mean absolute deviation

        acc_jerk.mean(),
        acc_jerk.max(),

        ax.std(), ay.std(), az.std(),
        ax.mean(), ay.mean(), az.mean(),
        np.mean(np.abs(ax) + np.abs(ay) + np.abs(az)),  # accel SMA

        gyro_mag.mean(),
        gyro_mag.std(),
        gyro_mag.max(),
        np.mean(gyro_mag ** 2),

        gx.std(), gy.std(), gz.std(),
        np.mean(np.abs(gx) + np.abs(gy) + np.abs(gz)),  # gyro SMA
    ]
    return np.array(feats, dtype=np.float64)


def extract_batch(windows: np.ndarray) -> np.ndarray:
    """windows: (N, WINDOW_SIZE, 6) -> (N, NUM_FEATURES)."""
    return np.stack([extract_features(w) for w in windows])
